x/bank/keeper/grpc_query_test.go

```diff
@@ -119,3 +120,173 @@ func (suite *IntegrationTestSuite) TestQueryParams() {
+			},
+			true,
+		},
+		{
+			"success, no results",
+			func() {
+				req = &types.QueryDenomsMetadataRequest{
+					Pagination: &query.PageRequest{
+						Limit:      3,
+						CountTotal: true,
+					},
+				}
+			},
+			true,
+		},
+		{
+			"success",
+			func() {
+				metadataAtom := types.Metadata{
+					Description: "The native staking token of the Cosmos Hub.",
+					DenomUnits: []*types.DenomUnit{
+						{
+							Denom:    "uatom",
+							Exponent: 0,
+							Aliases:  []string{"microatom"},
+						},
+						{
+							Denom:    "atom",
+							Exponent: 6,
+							Aliases:  []string{"ATOM"},
+						},
+					},
+					Base:    "uatom",
+					Display: "atom",
+				}
+
+				metadataEth := types.Metadata{
+					Description: "Ethereum native token",
+					DenomUnits: []*types.DenomUnit{
+						{
+							Denom:    "wei",
+							Exponent: 0,
+						},
+						{
+							Denom:    "eth",
+							Exponent: 18,
+							Aliases:  []string{"ETH", "ether"},
+						},
+					},
+					Base:    "wei",
+					Display: "eth",
+				}
+
+				suite.app.BankKeeper.SetDenomMetaData(suite.ctx, metadataAtom)
+				suite.app.BankKeeper.SetDenomMetaData(suite.ctx, metadataEth)
+				expMetadata = []types.Metadata{metadataAtom, metadataEth}
+				req = &types.QueryDenomsMetadataRequest{
+					Pagination: &query.PageRequest{
+						Limit:      7,
+						CountTotal: true,
+					},
+				}
+			},
+			true,
+		},
+	}
+
+	for _, tc := range testCases {
+		suite.Run(fmt.Sprintf("Case %s", tc.msg), func() {
+			suite.SetupTest() // reset
+
+			tc.malleate()
+			ctx := sdk.WrapSDKContext(suite.ctx)
+
+			res, err := suite.queryClient.DenomsMetadata(ctx, req)
+
+			if tc.expPass {
+				suite.Require().NoError(err)
+				suite.Require().NotNil(res)
+				suite.Require().Equal(expMetadata, res.Metadatas)
+			} else {
+				suite.Require().Error(err)
+			}
+		})
+	}
+}
+
+func (suite *IntegrationTestSuite) QueryDenomMetadataRequest() {
+	var (
+		req         *types.QueryDenomMetadataRequest
+		expMetadata = types.Metadata{}
+	)
+
+	testCases := []struct {
+		msg      string
+		malleate func()
+		expPass  bool
+	}{
+		{
+			"empty denom",
+			func() {
+				req = &types.QueryDenomMetadataRequest{}
+			},
+			false,
+		},
+		{
+			"success",
+			func() {
+				expMetadata := types.Metadata{
+					Description: "The native staking token of the Cosmos Hub.",
+					DenomUnits: []*types.DenomUnit{
+						{
+							Denom:    "uatom",
+							Exponent: 0,
+							Aliases:  []string{"microatom"},
+						},
+						{
+							Denom:    "atom",
+							Exponent: 6,
+							Aliases:  []string{"ATOM"},
+						},
+					},
+					Base:    "uatom",
+					Display: "atom",
+				}
+
+				suite.app.BankKeeper.SetDenomMetaData(suite.ctx, expMetadata)
+				req = &types.QueryDenomMetadataRequest{
+					Denom: expMetadata.Base,
+				}
+			},
+			true,
+		},
+	}
+
+	for _, tc := range testCases {
+		suite.Run(fmt.Sprintf("Case %s", tc.msg), func() {
+			suite.SetupTest() // reset
```

Comment:
Why calling this for every sub-test?

Reply:
to reset the request value

Reply:
It's to prevent leftover state from interfering with the expected results. If you call `SetDenomMetaData` and don't clear the stores, then the set metadata will carry over into the next test. This would make test case ordering matter

`tc.malleate()` will set the request value